File: android/src/main/kotlin/com/algorigo/in_app_updater/InAppUpdateInfo.kt
package com.algorigo.in_app_updater

import com.google.android.play.core.appupdate.AppUpdateInfo
import com.google.android.play.core.appupdate.AppUpdateOptions
import com.google.android.play.core.install.model.AppUpdateType
import com.google.android.play.core.install.model.InstallStatus
import com.google.android.play.core.install.model.UpdateAvailability
import com.google.android.play.core.ktx.bytesDownloaded
import com.google.android.play.core.ktx.installStatus
import com.google.android.play.core.ktx.isFlexibleUpdateAllowed
import com.google.android.play.core.ktx.isImmediateUpdateAllowed
import com.google.android.play.core.ktx.updatePriority

class InAppUpdateInfo(
  val appUpdateInfo: AppUpdateInfo
) {

  fun toMap() = mapOf(
    "updateAvailability" to appUpdateInfo.updateAvailability(),
    "availableVersionCode" to appUpdateInfo.availableVersionCode(),
    "updatePriority" to appUpdateInfo.updatePriority,
    "packageName" to appUpdateInfo.packageName(),
    "clientVersionStalenessDays" to appUpdateInfo.clientVersionStalenessDays(),
    "installStatus" to appUpdateInfo.installStatus,
    "isFlexibleUpdateAllowed" to isFlexibleUpdateAllowed,
    "isFlexibleUpdateFailedPreconditions" to appUpdateInfo
      .getFailedUpdatePreconditions(AppUpdateOptions.defaultOptions(AppUpdateType.FLEXIBLE))
      ?.map { it.toInt() }
      ?.toList(),
    "isImmediateUpdateAllowed" to isImmediateUpdateAllowed,
    "isImmediateUpdateFailedPreconditions" to appUpdateInfo
      .getFailedUpdatePreconditions(AppUpdateOptions.defaultOptions(AppUpdateType.IMMEDIATE))
      ?.map { it.toInt() }
      ?.toList(),
    "bytesDownloaded" to appUpdateInfo.bytesDownloaded(),
    "totalBytesToDownload" to appUpdateInfo.totalBytesToDownload(),
  )

  val isImmediateUpdateAllowed: Boolean
    get() = appUpdateInfo.isImmediateUpdateAllowed

  val isFlexibleUpdateAllowed: Boolean
    get() = appUpdateInfo.isFlexibleUpdateAllowed

  fun isUpdateAvailable(): Boolean = appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE

  fun isUpdateInProgress(): Boolean = appUpdateInfo.updateAvailability() == UpdateAvailability.DEVELOPER_TRIGGERED_UPDATE_IN_PROGRESS
}

File: lib/data/in_app_update_info.dart
import 'package:in_app_updater/data/update_availability.dart';
import 'package:in_app_updater/extensions/in_app_update_extensions.dart';

import 'install_status.dart';

class InAppUpdateInfo {
  final UpdateAvailability? updateAvailability;
  final int? availableVersionCode;
  final int? updatePriority;
  final String? packageName;
  final int? clientVersionStalenessDays;
  final InstallStatus? installStatus;
  final bool? isFlexibleUpdateAllowed;
  final List<int>? flexibleUpdateFailedPreconditions;
  final bool? isImmediateUpdateAllowed;
  final List<int>? immediateUpdateFailedPreconditions;
  final int? bytesDownloaded;
  final int? totalBytesToDownload;

  InAppUpdateInfo({
    required this.updateAvailability,
    required this.availableVersionCode,
    required this.updatePriority,
    required this.packageName,
    required this.clientVersionStalenessDays,
    required this.installStatus,
    required this.isFlexibleUpdateAllowed,
    required this.flexibleUpdateFailedPreconditions,
    required this.isImmediateUpdateAllowed,
    required this.immediateUpdateFailedPreconditions,
    required this.bytesDownloaded,
    required this.totalBytesToDownload,
  });

  factory InAppUpdateInfo.fromJson(Map<String, dynamic> json) {
    return InAppUpdateInfo(
      updateAvailability: (json['updateAvailability'] as int?)?.toUpdateAvailability(),
      availableVersionCode: json['availableVersionCode'],
      updatePriority: json['updatePriority'],
      packageName: json['packageName'],
      clientVersionStalenessDays: json['clientVersionStalenessDays'],
      installStatus: (json['installStatus'] as int?)?.toInstallStatus(),
      isFlexibleUpdateAllowed: json['isFlexibleUpdateAllowed'],
      flexibleUpdateFailedPreconditions: List<int>.from(json['isFlexibleUpdateFailedPreconditions']),
      isImmediateUpdateAllowed: json['isImmediateUpdateAllowed'],
      immediateUpdateFailedPreconditions: List<int>.from(json['isImmediateUpdateFailedPreconditions']),
      bytesDownloaded: json['bytesDownloaded'],
      totalBytesToDownload: json['totalBytesToDownload'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'updateAvailability': updateAvailability,
      'availableVersionCode': availableVersionCode,
      'updatePriority': updatePriority,
      'packageName': packageName,
      'clientVersionStalenessDays': clientVersionStalenessDays,
      'installStatus': installStatus,
      'isFlexibleUpdateAllowed': isFlexibleUpdateAllowed,
      'flexibleUpdateFailedPreconditions': flexibleUpdateFailedPreconditions,
      'isImmediateUpdateAllowed': isImmediateUpdateAllowed,
      'immediateUpdateFailedPreconditions': immediateUpdateFailedPreconditions,
      'bytesDownloaded': bytesDownloaded,
    };
  }
}
